File: app/javascript/template_builder/pdf_text_to_html.js
function escapeHtml (str) {
  return str
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#039;')
}

function isNumberedHeading (line) {
  return line.length <= 80 && /^\d+\.\s+[A-Z]/.test(line) && !/[.!?,;]$/.test(line)
}

function isAllCapsHeading (line) {
  return line.length >= 3 && !/[.!?,;]$/.test(line) && line === line.toUpperCase() && /[A-Z]/.test(line)
}

export function pdfTextToHtml (pageText) {
  if (!pageText) return ''

  const lines = pageText.split(/\r?\n/)
  let output = ''
  let inList = false

  for (const line of lines) {
    const stripped = line.trim()

    if (!stripped) {
      if (inList) { output += '</ul>'; inList = false }
      continue
    }

    if (isNumberedHeading(stripped)) {
      if (inList) { output += '</ul>'; inList = false }
      output += `<h3>${escapeHtml(stripped)}</h3>`
    } else if (isAllCapsHeading(stripped)) {
      if (inList) { output += '</ul>'; inList = false }
      output += `<h2>${escapeHtml(stripped)}</h2>`
    } else {
      const match = stripped.match(/^[•*-]\s+(.+)/)
      if (match) {
        if (!inList) { output += '<ul>'; inList = true }
        output += `<li>${escapeHtml(match[1])}</li>`
      } else {
        if (inList) { output += '</ul>'; inList = false }
        output += `<p dir="auto">${escapeHtml(stripped)}</p>`
      }
    }
  }

  if (inList) output += '</ul>'
  return output
}

File: lib/pdf_text_to_html.rb
# frozen_string_literal: true

module PdfTextToHtml
  module_function

  def call(page_text)
    output = +''
    current_list = nil

    page_text.split(/\r?\n/).each do |line|
      stripped = line.strip
      if stripped.empty?
        output << close_list(current_list) if current_list
        current_list = nil
        next
      end
      current_list = process_line(stripped, output, current_list)
    end

    output << close_list(current_list)
    output
  end

  def process_line(stripped, output, current_list)
    if numbered_heading?(stripped)
      output << close_list(current_list)
      output << "<h3>#{ERB::Util.html_escape(stripped)}</h3>"
      nil
    elsif all_caps_heading?(stripped)
      output << close_list(current_list)
      output << "<h2>#{ERB::Util.html_escape(stripped)}</h2>"
      nil
    elsif (match = stripped.match(/\A[•*-]\s+(.+)/))
      output << close_list(current_list) << '<ul>' unless current_list == :ul
      output << "<li>#{ERB::Util.html_escape(match[1])}</li>"
      :ul
    else
      output << close_list(current_list)
      output << %(<p dir="auto">#{ERB::Util.html_escape(stripped)}</p>)
      nil
    end
  end

  def numbered_heading?(line)
    line.length <= 80 && line.match?(/\A\d+\.\s+[A-Z]/) && !line.match?(/[.!?,;]\z/)
  end

  def all_caps_heading?(line)
    line.length >= 3 && !line.match?(/[.!?,;]\z/) &&
      line == line.upcase && line.match?(/[A-Z]/)
  end

  def close_list(current_list)
    case current_list
    when :ol then '</ol>'
    when :ul then '</ul>'
    else ''
    end
  end
end
